xwiki-commons-core/xwiki-commons-velocity/src/main/java/org/xwiki/velocity/internal/DefaultVelocityManager.java

```diff
@@ -0,0 +1,264 @@
+import org.xwiki.component.annotation.Component;
+import org.xwiki.component.manager.ComponentManager;
+import org.xwiki.component.phase.Initializable;
+import org.xwiki.component.phase.InitializationException;
+import org.xwiki.context.Execution;
+import org.xwiki.context.ExecutionContext;
+import org.xwiki.logging.LoggerConfiguration;
+import org.xwiki.script.ScriptContextManager;
+import org.xwiki.velocity.ScriptVelocityContext;
+import org.xwiki.velocity.VelocityConfiguration;
+import org.xwiki.velocity.VelocityEngine;
+import org.xwiki.velocity.VelocityManager;
+import org.xwiki.velocity.VelocityTemplate;
+import org.xwiki.velocity.XWikiVelocityException;
+import org.xwiki.velocity.internal.directive.TryCatchDirective;
+
+/**
+ * Note: This class should be moved to the Velocity module. However this is not possible right now since we need to
+ * populate the Velocity Context with XWiki objects that are located in the Core (such as the XWiki object for example)
+ * and since the Core needs to call the Velocity module this would cause a circular dependency.
+ *
+ * @version $Id$
+ * @since 15.9-rc-1
+ */
+@Component
+@Singleton
+public class DefaultVelocityManager implements VelocityManager, Initializable
+{
+    /**
+     * Used to access the current {@link org.xwiki.context.ExecutionContext}.
+     */
+    @Inject
+    protected Execution execution;
+
+    /**
+     * Used to get the current script context.
+     */
+    @Inject
+    private ScriptContextManager scriptContextManager;
+
+    @Inject
+    private LoggerConfiguration loggerConfiguration;
+
+    /**
+     * Velocity configuration to get the list of configured Velocity properties.
+     */
+    @Inject
+    private VelocityConfiguration velocityConfiguration;
+
+    @Inject
+    private ComponentManager componentManager;
+
+    @Inject
+    private Provider<InternalVelocityEngine> engineProvider;
+
+    /**
+     * Binding that should stay on Velocity side only.
+     */
+    private final Set<String> reservedBindings = new HashSet<>();
+
+    /**
+     * The actual Velocity runtime.
+     */
+    private RuntimeInstance runtimeInstance;
+
+    private VelocityEngine mainVelocityEngine;
+
+    @Override
+    public void initialize() throws InitializationException
+    {
+        // Set reserved bindings
+
+        // "context" is a reserved binding in JSR223 world
+        this.reservedBindings.add("context");
+
+        // Macros directive
+        this.reservedBindings.add("macro");
+        // Foreach directive
+        this.reservedBindings.add("foreach");
+        // Evaluate directive
+        this.reservedBindings.add("evaluate");
+        // TryCatch directive
+        this.reservedBindings.add("exception");
+        this.reservedBindings.add("try");
+        // Default directive
+        this.reservedBindings.add("define");
+        // The name of the context variable used for the template-level scope
+        this.reservedBindings.add("template");
```

Comment:
Just to be sure: there will never be a need to be able to add other custom reserved bindings? Wondering if those could be customized from a configuration property. I guess there will never be a need: if we need to add a binding it's because of an evolution of the velocity language no?

Reply:
Not that all that is nothing new. I just moved DefaultVelocityManager from platform (but without the macros.vm stuff).

Reply:
ok